pkg/volume/emptydir/empty_dir.go

```diff
@@ -421,6 +421,14 @@ func (ed *emptyDir) TearDown() error {
 
 // TearDownAt simply discards everything in the directory.
 func (ed *emptyDir) TearDownAt(dir string) error {
+	defer func() {
+		// remove ready dir which created in SetUp func
+		readyDir := ed.getMetaDir()
```

Comment:
Do you know why we need this ready file? I could not tell from the original pr.

Would the setup be more resilient if we also checked for the directory existing in addition to the ready file? We already check for tmpfs that it's a mount point.

Can you add a unit test for this new behavior?

Reply:
I have a theory based on digging old code and issues. https://github.com/kubernetes/kubernetes/pull/85136#issuecomment-676556803

We could add check on setup. Because setup should always create volume dir and then ready file, so logically it is enough to check ready file alone since it indicates volume dir already created. The issue is caused by teardown because it Only delete volume dir, not the ready file which leaves the situation. After this is fixed, then the logic should be right.

Reply:
at least changing permissions on empty directory or tmpfs mount should not take a long time. Maybe hugepages?

Doing only teardown means that we rely on teardown to be called in all cases. Do we handle reconstruction and uncertain mount scenarios properly?

Reply:
this is my guess, need to confirm with original author Paul.

Actually now checking other type of volumes which do not use this ready file, it might be some potential issues. e.g., in gce pd setup()

1. first setup call finishes mounting, and not yet to start to set ownership, now kubelet restarts
2. second setup call checks mount already exists, and will return success, without actually setting up ownership anymore?

ready files seems more useful for this type of volume type actually, which is not used..

Reply:
If we add the check of volum dir during setup, I think we should remove ready file First to avoid the following situation

1. in first setup call, because it creates volume dir and ready file, but fail to write payload, it will trigger teardown, volume dir is deleted, but ready file could not be removed temporally 
2. a second setup call is issued, check volume dir does not exist, create it and might not yet start to change ownership (takes some time), kubelet restarts
3. a third setup call comes, it see both ready file and volume dir, it will think it is ready and go ahead to write data (ownership is not set correctly yet?)

Reply:
If you remove ready file first during the teardown, does it solve that race?

I do agree if you need to call setup directories again, ready file should be removed first.

Reply:
For general issue of fsgroup, I do think it would be better to have it in two phases so we can recover from those partial scenarios. Related issues:
https://github.com/kubernetes/kubernetes/issues/78130
https://github.com/kubernetes/kubernetes/issues/89776

Reply:
The original ready file logic seems to have been introduced to indicate the emptyDir setup is complete. It doesn't seem to be related to the time taken for hugePages etc., At the time of [original CL](https://github.com/kubernetes/kubernetes/pull/9384/files), hugepageMedium support doesnt seem to exist. 

During teardown, if we remove the ready file first, then restarted kubelet might attempt to SetVolumeOwnership->changeFilePermission(). In that case, I am thinking if the current file permissions may prevent new permission changes.

Reply:
I think kubelet needs to run as root so that it can change any permissions

Reply:
Update: If current file permissions can prevent further permission changes, I think the current logic is no better. During setup, if kubelet crashes before creating readyfile, the restarted kubelet will reattempt to set the permissions.

Reply:
Yes, agree. running as root shouldn't prevent from setting the permissions.